README.md
# osu-parser

[![npm version](https://img.shields.io/npm/v/@brunohpaiva/osu-parser.svg?style=flat)](https://www.npmjs.org/package/@brunohpaiva/osu-parser)
[![npm downloads](https://img.shields.io/npm/dm/@brunohpaiva/osu-parser.svg?style=flat)](http://npm-stat.com/charts.html?package=@brunohpaiva/osu-parser)
[![npm bundle size](https://img.shields.io/bundlephobia/min/@brunohpaiva/osu-parser?style=flat)](https://www.npmjs.org/package/@brunohpaiva/osu-parser)

A library to read and write osu files for Node.js

## Installing

Using npm:

```bash
$ npm install @brunohpaiva/osu-parser
```

Using yarn:

```bash
$ yarn add @brunohpaiva/osu-parser
```

## Examples

#### Replay file (.osr)

Reading from a file
```ts
import { OsuReplay } from '@brunohpaiva/osu-parser';
import fs from 'fs';

// Path of local .osr file
const filePath = './osu-replay-input.osr';
// Read file contents synchronously and stores it in a buffer
const buffer = fs.readFileSync(filePath);
// Parses the buffer to a OsuReplay
const replay = OsuReplay.parse(buffer);

console.log(replay);
```

Writing to a file
```ts
import { OsuReplay } from '@brunohpaiva/osu-parser';
import fs from 'fs';

// Creates a new replay
const replay = new OsuReplay();

// Set some data to it
replay.type = 0; // osu! standard
replay.gameVersion = 20200724;
replay.beatmapHash = 'hash';
replay.playerName = 'player';
replay.replayHash = 'hash';
replay.count300s = 100;
replay.count100s = 50;
replay.count50s = 25;
replay.countGekis = 0;
replay.countKatus = 5;
replay.countMisses = 1;
replay.totalScore = 1953902;
replay.greatestCombo = 328;
replay.perfectCombo = false;
replay.modsUsed = ['DoubleTime'];
replay.lifeBarGraph = 'lifeBar';
replay.windowsTicks = 637321042198407465n;
replay.actions = [
  {
    timestamp: 10, // Time in milliseconds since previous action
    x: 10, // x coordinate from 0 to 512
    y: 23, // y coordinate from 0 to 384
    buttons: ['KeyboardOne'],
  },
  // More actions
];
replay.onlineScoreId = 98247527n;

const osuBuffer = replay.writeToOsuBuffer();
const buffer = osuBuffer.buffer;

// Path to store the generated .osr file
const filePath = './osu-replay-output.osr';
// Writes the buffer to a file synchronously
fs.writeFileSync(filePath, buffer);
```

#### OsuBuffer

Reading and writing with a `OsuBuffer`
```ts
import { OsuBuffer } from '@brunohpaiva/osu-parser';

// Initialize a empty buffer.
const osuBuffer = new OsuBuffer();
osuBuffer.writeInt(10);
osuBuffer.writeVarChar('A text');
osuBuffer.writeByte(5);

console.log(osuBuffer.buffer);
// Output: <Buffer 0a 00 00 00 0b 06 41 20 74 65 78 74 05>

console.log(osuBuffer.readInt());
// Output: 10

console.log(osuBuffer.readVarChar());
// Output: A text
```


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

### [0.2.3](https://github.com/brunohpaiva/osu-parser/compare/v0.2.2...v0.2.3) (2020-09-17)

### [0.2.2](https://github.com/brunohpaiva/osu-parser/compare/v0.2.1...v0.2.2) (2020-09-17)

### [0.2.1](https://github.com/brunohpaiva/osu-parser/compare/v0.2.0...v0.2.1) (2020-09-16)

## [0.2.0](https://github.com/brunohpaiva/osu-parser/compare/v0.1.6...v0.2.0) (2020-09-15)


### ⚠ BREAKING CHANGES

* Change in parameter order of the OsuBufferr's slice method.

### Features

* now using umireon's LZMA-JS implementation to (de)compress replay data ([f716ea8](https://github.com/brunohpaiva/osu-parser/commit/f716ea8b30adf7bcf04366b62b59b418690f2544))
* started replay parser ([93b0426](https://github.com/brunohpaiva/osu-parser/commit/93b04261489611a0fd9d6f7e20624c6cfa895676))
* the OsuBuffer's concat method now accepts a Buffer or OsuBuffer as the parameter ([c8e790d](https://github.com/brunohpaiva/osu-parser/commit/c8e790d18da178fa281f24ba427b99646bfcd908))


* changed parameters order in slice method and added default value for start ([8f09633](https://github.com/brunohpaiva/osu-parser/commit/8f0963392d45c9b8b1b87ce46b2031f34a24bff3))

### [0.1.6](https://github.com/brunohpaiva/osu-parser/compare/v0.1.5...v0.1.6) (2020-09-10)

### [0.1.5](https://github.com/brunohpaiva/osu-parser/compare/v0.1.4...v0.1.5) (2020-09-10)


### Features

* more read methods in OsuBuffer and basic documentation ([227820a](https://github.com/brunohpaiva/osu-parser/commit/227820ab82928cf3698ab65a1d5a8fed8d061949))

### 0.1.4 (2020-09-10)


src/lib/OsuBuffer.spec.ts
import test, { ExecutionContext, Macro } from 'ava';

import { OsuBuffer } from './OsuBuffer';

const createOsuBuffer = (...data: number[]) => new OsuBuffer(data);
const method = (prefix: string, type: string) =>
  `${prefix}${type.charAt(0).toUpperCase() + type.slice(1)}`;
type BufferType = number | bigint | string;

const readMacro: Macro<[string, BufferType, number[]]> = (
  t: ExecutionContext,
  type,
  expected,
  data
) => {
  t.is(createOsuBuffer(...data)[method('read', type)](), expected);
};

readMacro.title = (providedTitle, type) => `${providedTitle || type}: can read`;

const writeMacro: Macro<[string, BufferType]> = (
  t: ExecutionContext,
  type,
  expected
) => {
  const osuBuffer = createOsuBuffer();
  osuBuffer[method('write', type)](expected);
  t.is(osuBuffer[method('read', type)](), expected);
};

writeMacro.title = (providedTitle, type) =>
  `${providedTitle || type}: can write`;

test([readMacro, writeMacro], 'byte', 159, [0x9f]);
test([readMacro, writeMacro], 'short', 2827, [0x0b, 0x0b]);
test([readMacro, writeMacro], 'int', 671573433, [0xb9, 0x65, 0x07, 0x28]);
test([readMacro, writeMacro], 'long', 637321042198407465n, [
  0x29,
  0x29,
  0x1e,
  0x78,
  0x1c,
  0x38,
  0xd8,
  0x08,
]);

test('varInt: can read', (t) => {
  t.deepEqual(createOsuBuffer(0xe2, 0).readVarInt(), [98, 2]);
  t.deepEqual(createOsuBuffer(0, 0, 0xe2, 0).readVarInt(2), [98, 4]);
});

test('varInt: can write', (t) => {
  const buffer = createOsuBuffer();
  buffer.writeVarInt(98);
  t.deepEqual(buffer.readVarInt(), [98, 2]);
});

test('small varChar', [readMacro, writeMacro], 'varChar', 'Test', [
  0x0b,
  0x04,
  0x54,
  0x65,
  0x73,
  0x74,
]);

const LONG_VARCHAR =
  'Lorem ipsum dolor sit amet, consectetur adipiscing elit.' +
  'Nullam in nulla vel velit auctor pharetra.';

test('long varChar', [readMacro, writeMacro], 'varChar', LONG_VARCHAR, [
  0x0b,
  ...OsuBuffer.encodeVarInt(Buffer.byteLength(LONG_VARCHAR)),
  ...Buffer.from(LONG_VARCHAR),
]);

test('can write nullable empty varChar', (t) => {
  const osuBuffer = createOsuBuffer();
  osuBuffer.writeVarChar('', true);
  t.deepEqual(osuBuffer.buffer, Buffer.from([0]));
});

test('can write empty varChar', (t) => {
  const osuBuffer = createOsuBuffer();
  osuBuffer.writeVarChar('');
  t.deepEqual(osuBuffer.buffer, Buffer.from([0x0b, 0]));
});

test('can read empty varChar', (t) => {
  t.is(createOsuBuffer(0x0b, 0).readVarChar(), '');
});

test('can read varChar specifying the offset', (t) => {
  const osuBuffer = createOsuBuffer(0, 0, 0x0b, 0x04, 0x54, 0x65, 0x73, 0x74);
  t.is(osuBuffer.readVarChar(2), 'Test');
});

test('throws error when trying to encode varInt with value -1', (t) => {
  t.throws(() => OsuBuffer.encodeVarInt(-1), {
    message: /"value" is out of range/,
  });
});

test('throws error when reading a value that is not a varChar', (t) => {
  t.throws(
    () => createOsuBuffer(0x0a, 0x05, 0x54, 0x65, 0x73, 0x74).readVarChar(),
    {
      message: /Not a varChar/,
    }
  );
});

test('can create OsuBuffer from a Buffer', (t) => {
  const buffer = Buffer.from([
    0x0b,
    0x04,
    0x54,
    0x65,
    0x73,
    0x74,
    0x0b,
    0x0b,
    0x9f,
  ]);
  const osuBuffer = new OsuBuffer(buffer);
  t.is(osuBuffer.readVarChar(), 'Test');
  t.is(osuBuffer.readShort(), 2827);
  t.is(osuBuffer.readByte(), 159);
});

test('can create OsuBuffer from another OsuBuffer', (t) => {
  const inputOsuBuffer = createOsuBuffer(
    0x0b,
    0x04,
    0x54,
    0x65,
    0x73,
    0x74,
    0x0b,
    0x0b,
    0x9f
  );
  const osuBuffer = new OsuBuffer(inputOsuBuffer);
  t.is(osuBuffer.readVarChar(), inputOsuBuffer.readVarChar());
  t.is(osuBuffer.readShort(), inputOsuBuffer.readShort());
  t.is(osuBuffer.readByte(), inputOsuBuffer.readByte());
});


src/lib/utils/netTicks.spec.ts
import test from 'ava';

import { ticksFromDate, ticksToDate } from './netTicks';

test('can convert Date to .NET ticks', (t) => {
  const date = new Date('2020-08-04T02:16:59.840Z');
  t.is(ticksFromDate(date), 637321042198400000n);
});

test('can convert .NET ticks to Date', (t) => {
  t.deepEqual(
    ticksToDate(637321042198400000n),
    new Date('2020-08-04T02:16:59.840Z')
  );
});


src/lib/files/OsuReplay.spec.ts
import fs from 'fs';

import test from 'ava';

import { OsuReplay } from './OsuReplay';

test('able to write OsuReplay to buffer', (t) => {
  const replay = new OsuReplay();
  replay.type = 0;
  replay.gameVersion = 20200724;
  replay.beatmapHash = 'beatmaphash';
  replay.playerName = 'brunohpaiva';
  replay.replayHash = 'replayhash';
  replay.count300s = 261;
  replay.count100s = 7;
  replay.count50s = 0;
  replay.countGekis = 57;
  replay.countKatus = 5;
  replay.countMisses = 1;
  replay.totalScore = 1953902;
  replay.greatestCombo = 328;
  replay.perfectCombo = false;
  replay.modsUsed = ['DoubleTime'];
  replay.lifeBarGraph = 'lifeBar';
  replay.windowsTicks = 637321042198407465n;
  replay.actions = [
    {
      timestamp: 0,
      x: 10,
      y: 23,
      buttons: ['KeyboardOne'],
    },
    {
      timestamp: 3,
      x: 53,
      y: 200,
      buttons: ['KeyboardTwo', 'MouseTwo'],
    },
    {
      timestamp: -12345,
      x: 0,
      y: 0,
      rngSeed: 654321,
    },
  ];
  replay.onlineScoreId = 98247527n;

  const { buffer } = replay.writeToOsuBuffer();
  const expectedBuffer = fs.readFileSync(
    './test/osu-replay-output-expected.osr'
  );

  t.deepEqual(buffer, expectedBuffer);
});

test('able to parse OsuReplay from buffer', (t) => {
  const buffer = fs.readFileSync('./test/osu-replay-input.osr');
  const replay = OsuReplay.parse(buffer);

  t.like(replay, <OsuReplay>{
    type: 0,
    gameVersion: 20151228,
    beatmapHash: 'a5101f2dfa293807c894922b2876ea8d',
    playerName: 'Weabole',
    replayHash: 'b4f0ae8f74f810952f7677c180408d12',
    count300s: 264,
    count100s: 5,
    count50s: 0,
    countGekis: 59,
    countKatus: 4,
    countMisses: 0,
    totalScore: 3549720,
    greatestCombo: 396,
    perfectCombo: true,
    modsUsed: ['Hidden', 'HardRock', 'DoubleTime'],
    windowsTicks: 636897556980000000n,
    onlineScoreId: 2772099850n,
  });
  t.regex(replay.lifeBarGraph, /^(\d+\|[0-9.?]+,?)*$/);
  t.regex(replay.data, /^(-?\d+\|-?[0-9.]+\|-?[0-9.]+\|\d+,?){4658}$/);
  t.true(replay.actions.length === 4658); // 4657 actions + RNG seed
});

test('able to convert .NET ticks to date and vice-versa', (t) => {
  const replay = new OsuReplay();
  const date = new Date(2020, 8, 15, 0, 0, 0, 0);
  replay.date = date;
  t.is(replay.windowsTicks, 637357356000000000n);
  t.deepEqual(replay.date, date);
});


src/lib/OsuBuffer.ts
import { readBigUInt64LE, writeBigUInt64LE } from './utils';

/**
 * Class holding a {@link Buffer} that contains binary data from a osu file,
 * with utility methods to read and write values.
 */
export class OsuBuffer {
  /** Buffer holding all the binary data. */
  buffer: Buffer;
  /**
   * Zero-based indicating the current position used for reading from the Buffer.
   */
  position = 0;

  /**
   * Creates a OsuBuffer instance with optional initial data.
   * @param data Initial data.
   */
  constructor(data: OsuBuffer | Buffer | ArrayBuffer | number[] = []) {
    if (data instanceof OsuBuffer) {
      this.buffer = data.buffer;
    } else if (data instanceof Buffer) {
      this.buffer = data;
    } else {
      this.buffer = Buffer.from(data);
    }
  }

  /**
   * Concatenates a buffer to the current one.
   * @internal
   * @param otherBuffer A {@link Buffer} or {@link OsuBuffer} to append
   * to the {@link buffer current buffer}.
   */
  concat(otherBuffer: Buffer | OsuBuffer) {
    this.buffer = Buffer.concat([
      this.buffer,
      /* istanbul ignore next */
      otherBuffer instanceof Buffer ? otherBuffer : otherBuffer.buffer,
    ]);
  }

  /**
   * Slices the {@link buffer} returning a new one with the desired length.
   * @param length The length to slice.
   * @param start Zero-based start position to slice.
   * Default: {@link position current position} plus length parameter.
   * @returns A {@link Buffer} with size equals to the length parameter.
   */
  slice(length: number, start = this.positionIncrement(length)) {
    return this.buffer.slice(start, start + length);
  }

  /**
   * Reads a byte value from the {@link buffer}.
   * @param offset Zero-based position to start reading in.
   * Default: {@link position current position} plus one.
   * @returns The read byte.
   */
  readByte(offset = this.positionIncrement(1)) {
    return this.buffer.readUInt8(offset);
  }

  /**
   * Reads a short value from the {@link buffer}.
   * @param offset Zero-based position to start reading in.
   * Default: {@link position current position} plus two.
   * @returns The read short.
   */
  readShort(offset = this.positionIncrement(2)) {
    return this.buffer.readUInt16LE(offset);
  }

  /**
   * Reads a int value from the {@link buffer}.
   * @param offset Zero-based position to start reading in.
   * Default: {@link position current position} plus four.
   * @returns The read int.
   */
  readInt(offset = this.positionIncrement(4)) {
    return this.buffer.readUInt32LE(offset);
  }

  /**
   * Reads a long value from the {@link buffer}.
   * @param offset Zero-based position to start reading in.
   * Default: {@link position current position} plus eight.
   * @param useNodeImpl Boolean value representing when
   * to use the {@link Buffer#readBigUInt64LE NodeJS implementation}.
   * @returns The read long.
   */
  readLong(offset = this.positionIncrement(8), useNodeImpl?: boolean) {
    return readBigUInt64LE(this.buffer, offset, useNodeImpl);
  }

  /**
   * Reads a varInt value from the {@link buffer}.
   * @internal
   * @param startOffset start offset.
   * @returns A array containing first the varInt value,
   * and then a zero-based indicating where the varInt ends in the {@link buffer}.
   */
  readVarInt(startOffset?: number) {
    let offset =
      typeof startOffset !== 'undefined' ? startOffset : this.position;
    let result = 0;
    let shift = 0;
    // eslint-disable-next-line no-constant-condition
    while (true) {
      const byte = this.readByte(offset++);
      result |= (byte & 0x7f) << shift;
      shift += 7;
      if ((0x80 & byte) === 0) {
        /* istanbul ignore if */
        if (shift < 32 && (byte & 0x40) !== 0) {
          result = result | (~0 << shift);
          break;
        }
        break;
      }
    }

    if (typeof startOffset === 'undefined') {
      this.position = offset;
    }
    return [result, offset];
  }

  /**
   * Reads a varChar value from the {@link buffer}.
   * @param startOffset A.
   * @param encoding The varChar encoding to use when reading from the {@link buffer}.
   * Default: 'utf8'.
   * @returns The read varChar.
   */
  readVarChar(startOffset?: number, encoding?: BufferEncoding) {
    const suppliedStartOffset = typeof startOffset !== 'undefined';
    let offset = suppliedStartOffset ? startOffset : this.position;
    if (this.readByte(offset++) === 0x0b) {
      if (!suppliedStartOffset) this.position += 1;
      const [stringLength, endOffset] = this.readVarInt(offset);
      if (!suppliedStartOffset) this.position = endOffset + stringLength;
      return this.slice(stringLength, endOffset).toString(encoding);
    }
    throw new Error("The first byte isn't equals to 0x0b. Not a varChar.");
  }

  /**
   * Writes a int with a specified byte length.
   * @param value The value to write.
   * @param byteLength The length in bytes of the value.
   * @internal
   */
  private writeUInt(value: number, byteLength: number) {
    const buffer = Buffer.alloc(byteLength);
    buffer.writeUIntLE(value, 0, byteLength);
    this.concat(buffer);
  }

  /**
   * Writes a byte value to the {@link buffer}.
   * @param value The value to write.
   */
  writeByte(value: number) {
    this.writeUInt(value, 1);
  }

  /**
   * Writes a short value to the {@link buffer}.
   * @param value The value to write.
   */
  writeShort(value: number) {
    this.writeUInt(value, 2);
  }

  /**
   * Writes a int value to the {@link buffer}.
   * @param value The value to write.
   */
  writeInt(value: number) {
    this.writeUInt(value, 4);
  }

  /**
   * Writes a long value to the {@link buffer}.
   * @param value The value to write.
   * @param useNodeImpl Boolean value representing when
   * to use the {@link Buffer#writeBigUInt64LE NodeJS implementation}.
   */
  writeLong(value: bigint, useNodeImpl?: boolean) {
    const buffer = Buffer.alloc(8);
    writeBigUInt64LE(buffer, value, 0, useNodeImpl);
    this.concat(buffer);
  }

  /**
   * Encodes a number to a varInt.
   * @param value The value to encode.
   * @returns The encoded value as a array of bytes.
   */
  static encodeVarInt(value: number): number[] {
    if (value < 0) {
      throw new RangeError(
        `The value of "value" is out of range. It must be >= 0`
      );
    }

    value |= 0;
    const bytes = [];
    // eslint-disable-next-line no-constant-condition
    while (true) {
      const byte = value & 0x7f;
      value >>= 7;
      if (value === 0 && (byte & 0x40) === 0) {
        bytes.push(byte);
        break;
      }
      bytes.push(byte | 0x80);
    }
    return bytes;
  }

  /**
   * Writes a varInt value to the {@link buffer}.
   * @param value The value to write.
   */
  writeVarInt(value: number) {
    this.concat(Buffer.from(OsuBuffer.encodeVarInt(value)));
  }

  /**
   * Writes a varChar value to the {@link buffer}.
   * @param value The value to write.
   * @param nullable Boolean representing if the varChar is nullable.
   * @param encoding The varChar encoding to use when writing to the {@link buffer}.
   * Default: 'utf8'.
   */
  writeVarChar(value: string, nullable = false, encoding?: BufferEncoding) {
    const bytes = [];
    if (typeof value === 'undefined' || value.length === 0) {
      if (nullable) {
        bytes.push(0);
      } else {
        bytes.push(0x0b, 0);
      }
    } else {
      const length = Buffer.byteLength(value, encoding);
      bytes.push(0x0b, ...OsuBuffer.encodeVarInt(length));

      const stringBuffer = Buffer.from(value, encoding);
      bytes.push(...stringBuffer);
    }

    this.concat(Buffer.from(bytes));
  }

  /**
   * Adds the length to the {@link position current position} variable.
   * @internal
   * @param length The length to increment in {@link position}.
   * @returns The old position.
   */
  positionIncrement(length: number) {
    const position = this.position;
    this.position += length;
    return position;
  }
}


src/lib/utils/netTicks.ts
/**
 * Time in milliseconds since 12:00:00 midnight, January 1, 0001.
 * @internal
 */
const NET_START_DATE = new Date('0001-01-01').getTime();
/**
 * {@link NET_START_DATE} in seconds.
 * @internal
 */
const SECONDS_TO_EPOCH = -NET_START_DATE / 1000;
/**
 * Number of ticks in a millisecond.
 * @internal
 */
const TICKS_PER_MS = 10000;
/**
 * Number of milliseconds in a second.
 * @internal
 */
const MS_PER_SEC = 1000;

/**
 * Converts .NET ticks to a {@link Date}.
 * @internal
 * @param value Number of .NET ticks to convert.
 * @returns The date representing the .NET ticks.
 */
export function ticksToDate(value: bigint) {
  return new Date(
    Number((value * 100n) / BigInt(TICKS_PER_MS)) / 100 + NET_START_DATE
  );
}

/**
 * Converts milliseconds to .NET ticks.
 * @internal
 * @param value Number of milliseconds to convert.
 * @returns Number of .NET ticks representing the specified milliseconds.
 */
export function ticksFromMilliseconds(value: number) {
  return BigInt((value + SECONDS_TO_EPOCH * MS_PER_SEC) * TICKS_PER_MS);
}

/**
 * Converts a {@link Date} to .NET ticks.
 * @internal
 * @param value The date to convert.
 * @returns Number of .NET ticks representing the specified {@link Date}.
 */
export function ticksFromDate(value: Date) {
  return ticksFromMilliseconds(value.getTime());
}


src/lib/utils/readBigUInt64LE.ts
/**
 * Reads a BigInt number from a Buffer.
 * @internal
 * @param buffer The source buffer.
 * @param offset Zero-based index number indicating the start position to
 * start reading on. Default: 0
 * @param useNodeImpl Boolean value representing when
 * to use the {@link Buffer#readBigUInt64LE NodeJS implementation}.
 */
export function readBigUInt64LE(
  buffer: Buffer,
  offset = 0,
  useNodeImpl = typeof buffer.readBigUInt64LE !== 'undefined'
): bigint {
  if (useNodeImpl) {
    return buffer.readBigUInt64LE(offset);
  }

  if (buffer[offset] === undefined || buffer[offset + 7] === undefined)
    throw new RangeError(
      `The value of "offset" is out of range. It must be >= 0 and <= ${
        buffer.length - 8
      }. Received ${offset}`
    );

  const first = buffer[offset];
  const last = buffer[offset + 7];
  const lo =
    first +
    buffer[++offset] * 2 ** 8 +
    buffer[++offset] * 2 ** 16 +
    buffer[++offset] * 2 ** 24;

  const hi =
    buffer[++offset] +
    buffer[++offset] * 2 ** 8 +
    buffer[++offset] * 2 ** 16 +
    last * 2 ** 24;

  return BigInt(lo) + (BigInt(hi) << 32n);
}


src/lib/files/OsuReplay.ts
import { compress, decompress } from 'lzma';

import { OsuBuffer } from '../OsuBuffer';
import { ticksFromDate, ticksToDate } from '../utils';

/**
 * Enum holding all osu gameplay keys and bit values.
 */
export enum OsuButtonsEnum {
  MouseOne = 1,
  MouseTwo = 2,
  KeyboardOne = 4,
  KeyboardTwo = 8,
  Smoke = 16,
}

/**
 * Type representing a osu gameplay key.
 */
export type OsuButton = keyof typeof OsuButtonsEnum;

/**
 * Interface that represents a osu action.
 */
export interface OsuAction {
  timestamp: number;
  x: number;
  y: number;
  buttons?: OsuButton[];
  rngSeed?: number;
}

/**
 * Enum holding all osu mod names and bit values.
 */
export enum OsuModsEnum {
  'NoFail' = 1,
  'Easy' = 2,
  'TouchDevice' = 4,
  'Hidden' = 8,
  'HardRock' = 16,
  'SuddenDeath' = 32,
  'DoubleTime' = 64,
  'Relax' = 128,
  'HalfTime' = 256,
  'Nightcore' = 512,
  'Flashlight' = 1024,
  'Autoplay' = 2048,
  'SpunOut' = 4096,
  'Relax2' = 8192,
  'Perfect' = 16384,
  'Key4' = 32768,
  'Key5' = 65536,
  'Key6' = 131072,
  'Key7' = 262144,
  'Key8' = 524288,
  'FadeIn' = 1048576,
  'Random' = 2097152,
  'Cinema' = 4194304,
  'Target' = 8388608,
  'Key9' = 16777216,
  'KeyCoop' = 33554432,
  'Key1' = 67108864,
  'Key3' = 134217728,
  'Key2' = 268435456,
  'ScoreV2' = 536870912,
  'Mirror' = 1073741824,
  'KeyMod' = Key1 |
    Key2 |
    Key3 |
    Key4 |
    Key5 |
    Key6 |
    Key7 |
    Key8 |
    Key9 |
    KeyCoop,
  'FreeModAllowed' = NoFail |
    Easy |
    Hidden |
    HardRock |
    SuddenDeath |
    Flashlight |
    FadeIn |
    Relax |
    Relax2 |
    SpunOut |
    KeyMod,
  'ScoreIncreaseMods' = Hidden | HardRock | DoubleTime | Flashlight | FadeIn,
}

/**
 * Type representing a osu mod. Examples: 'Perfect', 'Easy', 'NoFail'
 */
export type OsuMod = keyof typeof OsuModsEnum;

/**
 * Class representing a
 * [osu replay file](https://osu.ppy.sh/help/wiki/osu%21_File_Formats/Osr_%28file_format%29).
 */
export class OsuReplay {
  /**
   * Game mode of the replay. Default: 0
   * ```
   * 0 = osu! Standard
   * 1 = Taiko
   * 2 = Catch the Beat
   * 3 = osu!mania.
   * ```
   */
  type = 0;
  /** Version of the game when the replay was created (ex. 20131216). Default: 0 */
  gameVersion = 0;
  /** Beatmap MD5 hash. Default: '' */
  beatmapHash = '';
  /** Player name. Default: '' */
  playerName = '';
  /** Replay MD5 hash (includes certain properties of the replay). Default: '' */
  replayHash = '';
  /** Number of 300s. Default: 0 */
  count300s = 0;
  /**
   * Number of 100s in standard, 150s in Taiko, 100s in CTB, 100s in mania.
   * Default: 0
   */
  count100s = 0;
  /** Number of 50s in standard, small fruit in CTB, 50s in mania. Default: 0 */
  count50s = 0;
  /** Number of Gekis in standard, Max 300s in mania. Default: 0 */
  countGekis = 0;
  /** Number of Katus in standard, 200s in mania. Default: 0 */
  countKatus = 0;
  /** Number of misses. Default: 0 */
  countMisses = 0;
  /** Total score displayed on the score report. Default: 0 */
  totalScore = 0;
  /** Greatest combo displayed on the score report. Default: 0 */
  greatestCombo = 0;
  /**
   * If it was a perfect/full combo. Only true with no misses, no slider breaks
   * and no early finished sliders. Default: false
   */
  perfectCombo = false;
  /** Mods used in the gameplay. Default: [] */
  modsUsed: OsuMod[] = [];
  /** Life bar graph. Default: '' */
  lifeBarGraph = '';
  /**
   * Time stamp ([Windows ticks](http://msdn.microsoft.com/en-us/library/system.datetime.ticks%28v=vs.110%29.aspx))
   * Default: 0n
   */
  windowsTicks = 0n;
  /** Serialized actions data separated by commas. Default: '' */
  private _data = '';
  /** Online Score ID. Default: 0n */
  onlineScoreId = 0n;
  /**
   * {@link data} parsed as a array of {@link OsuAction}.
   */
  actions: OsuAction[] = [];

  /**
   * Converts the {@link windowsTicks} to a {@link Date}.
   * @returns A {@link Date} instance representing the value of {@link windowsTicks}.
   */
  get date() {
    return ticksToDate(this.windowsTicks);
  }

  /**
   * Sets a new {@link windowsTicks} value from a {@link Date}.
   * @param date The new {@link windowsTicks} value to set represented as a {@link Date}.
   */
  set date(date: Date) {
    this.windowsTicks = ticksFromDate(date);
  }

  /**
   * Serialized actions data separated by commas.
   * @returns The serialized data that was parsed from the parsed buffer.
   */
  get data() {
    return this._data;
  }

  /**
   * Parses the serialized actions string to a {@link OsuAction} array.
   * @param serializedActions The string with the serialized actions.
   * @returns A array of {@link OsuAction}.
   */
  static unserializeActions(serializedActions: string): OsuAction[] {
    return serializedActions
      .split(',')
      .filter((string) => string.includes('|'))
      .map<OsuAction>((string) => {
        const actionData = string.split('|');
        const action = {
          timestamp: parseInt(actionData[0]),
          x: parseFloat(actionData[1]),
          y: parseFloat(actionData[2]),
        } as OsuAction;

        const number = parseInt(actionData[3]);

        if (!string.startsWith('-12345')) {
          action.buttons = Object.keys(OsuButtonsEnum)
            .filter((k: string | OsuButton) => {
              if (typeof OsuButtonsEnum[k] === 'string') {
                const bit = parseInt(k);
                return bit === (number & bit);
              }
              return false;
            })
            .map<OsuButton>((k: string) => OsuButtonsEnum[k]);
        } else {
          action.rngSeed = number;
        }

        return action;
      });
  }

  /**
   * Serializes a array of {@link OsuAction} into a string.
   * @param actions The array of {@link OsuAction} to serialize.
   * @returns The serialized actions as a string.
   */
  static serializeActions(actions: OsuAction[]) {
    return actions
      .map<string>((action) => {
        const isRngSeed =
          typeof action.buttons === 'undefined' &&
          typeof action.rngSeed !== 'undefined';
        if (isRngSeed) {
          return `-12345|0|0|${action.rngSeed}`;
        } else {
          const bitwise = action.buttons.reduce<number>(
            (prev, button) => prev + OsuButtonsEnum[button],
            0
          );
          return `${action.timestamp}|${action.x}|${action.y}|${bitwise}`;
        }
      })
      .join(',');
  }

  /**
   * Parses data of a osu replay file from a source.
   * @param source Source that represents a osu replay file.
   * @returns The parsed {@link OsuReplay}.
   */
  static parse(source: OsuBuffer | Buffer | ArrayBuffer | number[]) {
    const buffer =
      source instanceof OsuBuffer
        ? /* istanbul ignore next */ source
        : new OsuBuffer(source);
    buffer.position = 0;

    const replay = new OsuReplay();
    replay.type = buffer.readByte();
    replay.gameVersion = buffer.readInt();
    replay.beatmapHash = buffer.readVarChar();
    replay.playerName = buffer.readVarChar();
    replay.replayHash = buffer.readVarChar();
    replay.count300s = buffer.readShort();
    replay.count100s = buffer.readShort();
    replay.count50s = buffer.readShort();
    replay.countGekis = buffer.readShort();
    replay.countKatus = buffer.readShort();
    replay.countMisses = buffer.readShort();
    replay.totalScore = buffer.readInt();
    replay.greatestCombo = buffer.readShort();
    replay.perfectCombo = Boolean(buffer.readByte());

    let modsBitwise = buffer.readInt();
    const mods: OsuMod[] = [];
    while (modsBitwise >= 1) {
      const modBits = modsBitwise & (~modsBitwise + 1);
      mods.push(OsuModsEnum[modBits] as OsuMod);
      modsBitwise ^= modBits;
    }

    replay.modsUsed = mods;
    replay.lifeBarGraph = buffer.readVarChar();
    replay.windowsTicks = buffer.readLong();

    const compressedDataLength = buffer.readInt();
    const compressedData = buffer.slice(compressedDataLength);
    const uncompressedData = decompress(compressedData);
    replay._data = uncompressedData;
    replay.actions = OsuReplay.unserializeActions(uncompressedData);

    replay.onlineScoreId = buffer.readLong();

    return replay;
  }

  /**
   * Writes the replay data in this class instance to a {@link OsuBuffer}.
   * @returns The {@link OsuBuffer} holding the replay data.
   */
  writeToOsuBuffer() {
    const buffer = new OsuBuffer();
    buffer.writeByte(this.type);
    buffer.writeInt(this.gameVersion);
    buffer.writeVarChar(this.beatmapHash);
    buffer.writeVarChar(this.playerName);
    buffer.writeVarChar(this.replayHash);
    buffer.writeShort(this.count300s);
    buffer.writeShort(this.count100s);
    buffer.writeShort(this.count50s);
    buffer.writeShort(this.countGekis);
    buffer.writeShort(this.countKatus);
    buffer.writeShort(this.countMisses);
    buffer.writeInt(this.totalScore);
    buffer.writeShort(this.greatestCombo);
    buffer.writeByte(/* istanbul ignore next */ this.perfectCombo ? 1 : 0);
    buffer.writeInt(
      this.modsUsed.reduce<number>((int, mod) => int + OsuModsEnum[mod], 0)
    );
    buffer.writeVarChar(this.lifeBarGraph);
    buffer.writeLong(this.windowsTicks);

    const data = OsuReplay.serializeActions(this.actions);
    const compressedData = compress(data);
    buffer.writeInt(compressedData.length);
    buffer.concat(Buffer.from(compressedData));

    buffer.writeLong(this.onlineScoreId);
    return buffer;
  }
}
